ssl/t1_lib.c

```diff
@@ -1241,6 +1241,12 @@ TICKET_RETURN tls_get_ticket_from_client(SSL *s, CLIENTHELLO_MSG *hello,
 
     retv = tls_decrypt_ticket(s, PACKET_data(&ticketext->data), size,
                               hello->session_id, hello->session_id_len, ret);
+
+    if (s->session_ctx->decrypt_ticket_cb != NULL)
+        retv = s->session_ctx->decrypt_ticket_cb(s, *ret,
+                                                 PACKET_data(&ticketext->data),
```

Comment:
The name of this arg in the typedef suggests you are trying to send the ticket key name. If so we haven't sent any length of the data through as part of the callback. If we are not going to do that then we will need to validate that there are at least `TLSEXT_KEYNAME_LENGTH` bytes in the `PACKET`. We also don't expose `TLSEXT_KEYNAME_LENGTH` anywhere in a public header.

Reply:
`tls_decrypt_ticket()` will fail with SSL_TICKET_NO_DECRYPT if there isn't enough remaining data, but that return value isn't checked here. Yes, I will need to pass in TLSEXT_KEYNAME_LENGTH as an argument (or publicize it, which I'd rather not do).

Reply:
The callback must check the value of ret to determine what it can do. Updated documentation to reflect that

Reply:
> tls_decrypt_ticket() will fail with SSL_TICKET_NO_DECRYPT if there isn't enough remaining data, but that return value isn't checked here

If we are relying on that then a comment is probably appropriate. Or (possibly) even an assert/ossl_assert.

Reply:
> The callback must check the value of ret to determine what it can do. Updated documentation to reflect that

I kind of think that the keyname length should be no greater than the amount of data in the packet, i.e. if PACKET_remaining(&ticketext) < TLSEXT_KEYNAME_LENGTH then either:

1) We should pass NULL for the keyname and 0 for the length
or
2) We should pass the pointer into the packet where we expect the keyname, and the amount of data in the packet for the length.

Thinking about it though, if we don't have a valid key name then perhaps (1) is the safer option (what can the application usefully do with an invalid keyname?).

I also think the documentation needs to be a bit more explicit in saying that it gets called even in failure scenarios, so you might not have a valid ticket or a valid keyname.

Finally can you explain why keyname needs to be passed at all?